velox/type/tests/TimestampTest.cpp

```diff
@@ -299,37 +299,129 @@ TEST(TimestampTest, outOfRange) {
+    SCOPED_TRACE(fmt::format("epoch={}", epoch));
+    if (gmtime_r(&epoch, &expected)) {
+      ASSERT_TRUE(Timestamp::epochToUtc(epoch, actual));
+      checkTm(actual, expected);
+    } else {
+      ASSERT_FALSE(Timestamp::epochToUtc(epoch, actual));
+    }
+  }
+}
+
+void testTmToString(
+    const std::string& format,
+    const TimestampToStringOptions& options) {
+  uint64_t seed = 42;
+  std::default_random_engine gen(seed);
+  std::default_random_engine nanosGen(seed);
+
+  std::uniform_int_distribution<time_t> dist(
+      std::numeric_limits<time_t>::min(), std::numeric_limits<time_t>::max());
+  std::uniform_int_distribution<int> nanosDist(0, Timestamp::kMaxNanos);
+
+  std::tm actual{};
+  std::tm expected{};
+
+  for (int i = 0; i < 10'000; ++i) {
     auto epoch = dist(gen);
+    auto nanos = nanosDist(nanosGen);
     SCOPED_TRACE(fmt::format("epoch={}", epoch));
     if (gmtime_r(&epoch, &expected)) {
-      ASSERT_TRUE(epochToUtc(epoch, actual));
-      ASSERT_EQ(expected.tm_year, actual.tm_year);
-      ASSERT_EQ(expected.tm_yday, actual.tm_yday);
-      ASSERT_EQ(expected.tm_mon, actual.tm_mon);
-      ASSERT_EQ(expected.tm_mday, actual.tm_mday);
-      ASSERT_EQ(expected.tm_wday, actual.tm_wday);
-      ASSERT_EQ(expected.tm_hour, actual.tm_hour);
-      ASSERT_EQ(expected.tm_min, actual.tm_min);
-      ASSERT_EQ(expected.tm_sec, actual.tm_sec);
+      ASSERT_TRUE(Timestamp::epochToUtc(epoch, actual));
+      checkTm(actual, expected);
+
+      auto actualMillisStr = Timestamp::tmToString(actual, nanos, options);
+      auto expectedMillisStr = tmToString(expected, nanos, format, options);
+      ASSERT_EQ(expectedMillisStr, actualMillisStr);
+
     } else {
-      ASSERT_FALSE(epochToUtc(epoch, actual));
+      ASSERT_FALSE(Timestamp::epochToUtc(epoch, actual));
     }
   }
 }
 
+TEST(TimestampTest, tmToStringDateOnly) {
+  TimestampToStringOptions options;
+  options.mode = TimestampToStringOptions::Mode::kDateOnly;
+  testTmToString("%F", options);
+}
+
+TEST(TimestampTest, tmToStringTimeOnly) {
+  TimestampToStringOptions options;
+  options.mode = TimestampToStringOptions::Mode::kTimeOnly;
+  options.precision = TimestampToStringOptions::Precision::kMilliseconds;
+  testTmToString("%T", options);
+
+  options.precision = TimestampToStringOptions::Precision::kNanoseconds;
+  testTmToString("%T", options);
+}
+
+TEST(TimestampTest, tmToStringTimestamp) {
+  TimestampToStringOptions options;
+  options.mode = TimestampToStringOptions::Mode::kFull;
+  options.precision = TimestampToStringOptions::Precision::kMilliseconds;
+  testTmToString("%FT%T", options);
+
+  options.precision = TimestampToStringOptions::Precision::kNanoseconds;
+  testTmToString("%FT%T", options);
+}
+
+TEST(TimestampTest, tmToStringTimestampWithDefaultMode) {
+  // Default Mode
+  TimestampToStringOptions options;
+  options.precision = TimestampToStringOptions::Precision::kMilliseconds;
+  testTmToString("%FT%T", options);
```

Comment:
Since all test cases call actualMillisStr twice with different values for precision, consider extending testTmToString to test both values of precision to reduce copy-paste.

Reply:
Also, consider passing just TimestampToStringOptions::Mode to testTmToString since this is the only difference between all these test cases.

Reply:
@mbasmanova Refactored, please help me review this again, thank you.